### PR TITLE
fs: Various cleanups & add path class for later use

### PR DESCRIPTION
**Pull request by FearlessTobi**
This PR makes numerous changes to the fs system, including moving files, unifying error code naming and making fs definitions more consistent with LibHac and other sources.
It also adds the path class for use in future code additions.

I highly advise to review commit-by-commit.

### CODE REVIEW

File: src/core/hle/service/filesystem/fsp/fsp_ldr.h
Comment (liamwhite) [inline src/core/hle/service/filesystem/fsp/fsp_ldr.h:1]: This commit `fs: Move additional files to the fsp directory` should be merged into the previous commit `fs: Move fsp_srv subclasses to separate files`

File: src/core/file_sys/fs_file.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+namespace FileSys {
+
+struct ReadOption {
+    u32 _value;
```
Comment (liamwhite) [inline src/core/file_sys/fs_file.h:9]: Identifier names should not have leading underscores

File: src/core/file_sys/fs_file.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+static_assert(sizeof(ReadOption) == sizeof(u32));
+
+enum WriteOptionFlag : u32 {
+    WriteOptionFlag_None = (0 << 0),
+    WriteOptionFlag_Flush = (1 << 0),
+};
+
+struct WriteOption {
+    u32 _value;
```
Comment (liamwhite) [inline src/core/file_sys/fs_file.h:36]: Identifier names should not have leading underscores

File: src/core/file_sys/fs_memory_management.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <mutex>
+#include "common/alignment.h"
+
+namespace FileSys {
+
+std::mutex g_mutex;
```
Comment (liamwhite) [inline src/core/file_sys/fs_memory_management.h:11]: This mutex is not needed

File: src/core/file_sys/fs_memory_management.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+constexpr size_t RequiredAlignment = alignof(u64);
+
+void* AllocateUnsafe(size_t size) {
+    /* Allocate. */
```
Comment (liamwhite) [inline src/core/file_sys/fs_memory_management.h:16]: `/* */` C-style comments should be changed to `//` C++-style comments

File: src/core/file_sys/fs_file.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```
Comment (liamwhite) [inline src/core/file_sys/fs_file.h:1]: Outdated copyright year

File: src/core/file_sys/fs_memory_management.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```
Comment (liamwhite) [inline src/core/file_sys/fs_memory_management.h:1]: Outdated copyright year

File: src/core/file_sys/fs_operate_range.h [hunk 7]
```diff
@@ -0,0 +1,20 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```
Comment (liamwhite) [inline src/core/file_sys/fs_operate_range.h:1]: Outdated copyright year

File: src/core/file_sys/fs_path.h [hunk 8]
```diff
@@ -0,0 +1,570 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```
Comment (liamwhite) [inline src/core/file_sys/fs_path.h:1]: Outdated copyright year

File: src/core/file_sys/fs_path.h [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+    class WriteBuffer {
+        YUZU_NON_COPYABLE(WriteBuffer);
+
+    private:
+        char* m_buffer;
+        size_t m_length_and_is_normalized;
+
+    public:
+        constexpr WriteBuffer() : m_buffer(nullptr), m_length_and_is_normalized(0) { /* ... */
+        }
```
Comment (liamwhite) [inline src/core/file_sys/fs_path.h:38]: `constexpr WriteBuffer() : m_buffer(nullptr), m_length_and_is_normalized(0) {}`

File: src/core/file_sys/fs_path.h [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+    };
+
+private:
+    const char* m_str;
+    WriteBuffer m_write_buffer;
+
+public:
+    constexpr Path() : m_str(EmptyPath), m_write_buffer() {
+        /* ... */
+    }
```
Comment (liamwhite) [inline src/core/file_sys/fs_path.h:118]: `constexpr Path() : m_str(EmptyPath), m_write_buffer() {}`

File: src/core/file_sys/fs_path.h [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+
+    constexpr Path(const char* s) : m_str(s), m_write_buffer() {
+        m_write_buffer.SetNormalized();
+    }
+
+    constexpr ~Path() { /* ... */
+    }
```
Comment (liamwhite) [inline src/core/file_sys/fs_path.h:125]: `constexpr ~Path() = default;`

File: src/core/file_sys/fs_path.h [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+
+    constexpr Result SetShallowBuffer(const char* buffer) {
+        /* Check pre-conditions. */
```
Comment (liamwhite) [inline src/core/file_sys/fs_path.h:128]: `/* */` C-style comments should be changed to `//` C++-style comments

File: src/core/file_sys/fs_path_utility.h [hunk 13]
```diff
@@ -0,0 +1,1240 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```
Comment (liamwhite) [inline src/core/file_sys/fs_path_utility.h:1]: Copyright year

File: src/core/file_sys/fs_path_utility.h [hunk 14]
```diff
@@ -0,0 +1,1240 @@
...
+#include "common/scope_exit.h"
+#include "core/file_sys/fs_directory.h"
+#include "core/file_sys/fs_memory_management.h"
+#include "core/file_sys/fs_string_util.h"
+#include "core/hle/result.h"
+
+namespace FileSys {
+
+constexpr inline size_t MountNameLengthMax = 15;
+using Result = ::Result;
```
Comment (liamwhite) [inline src/core/file_sys/fs_path_utility.h:17]: Is this needed?

File: src/core/file_sys/fs_path_utility.h [hunk 15]
```diff
@@ -0,0 +1,1240 @@
...
+    return path[0] == StringTraits::Dot && path[1] == StringTraits::Dot &&
+           (path[2] == StringTraits::NullTerminator || path[2] == StringTraits::DirectorySeparator);
+}
+
+constexpr inline bool IsPathStartWithCurrentDirectory(const char* path) {
+    return IsCurrentDirectory(path) || IsParentDirectory(path);
+}
+
+constexpr inline bool IsSubPath(const char* lhs, const char* rhs) {
+    /* Check pre-conditions. */
```
Comment (liamwhite) [inline src/core/file_sys/fs_path_utility.h:180]: C-style comments

File: src/core/file_sys/fs_string_util.h [hunk 16]
```diff
@@ -0,0 +1,241 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```
Comment (liamwhite) [inline src/core/file_sys/fs_string_util.h:1]: Copyright year

File: src/core/file_sys/fs_string_util.h [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    while (*cur) {
+        cur++;
+    }
+
+    return static_cast<int>(cur - src);
+}
+
+/* std::size() does not support zero-size C arrays. We're fixing that. */
```
Comment (liamwhite) [inline src/core/file_sys/fs_string_util.h:74]: This is Common::Size (in common_funcs.h)

File: src/core/file_sys/fs_string_util.h [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+
+    static constexpr char GetUtf8NBytes(size_t i) {
+        return static_cast<char>(Utf8NBytesInnerTable[1 + i]);
+    }
+};
+
+} // namespace impl
+
+constexpr inline CharacterEncodingResult ConvertCharacterUtf8ToUtf32(u32* dst, const char* src) {
+    /* Check pre-conditions. */
```
Comment (liamwhite) [inline src/core/file_sys/fs_string_util.h:119]: C-style comments

**Review by liamwhite — APPROVED** [review]

**Comment by FearlessTobi** [comment]
Rebased.

**Pull request merged**